### PR TITLE
Use package name in GlobalKey hash

### PR DESCRIPTION
**Pull request by simonmaxen-da**
The daml repository has been updated to use package name based globalkey hash construction.

Until canton has been updated to use package based key construction the daml a tactical change has been made to call the shared key based constructor from the package based constructor. This will be reversed once canton is updated.

Removed test-only `Node.updateVersion` (as this now also needs a package name if updating the version of `Dev`). Have replaced with test-only `NodeVersionUpdater`

### CODE REVIEW

File: daml-lf/api-type-signature/src/main/scala/lf/typesig/PackageSignature.scala [hunk 1]
```diff
@@ -233,9 +233,9 @@ object PackageSignature {
     }
 
     (0 until newSignatures.size).foldLeft((s, newSignatures)) {
-      case (st @ (_, newSignatures), ifcK) =>
+      case (st @ (_, newSignatures1), ifcK) =>
         val ((s2, newSignatures2), newAtIfcK) =
-          newSignatures(ifcK).resolveRetroImplements(st)(findTpl)
+          newSignatures1(ifcK).resolveRetroImplements(st)(findTpl)
```
Comment (remyhaemmerle-da) [inline daml-lf/api-type-signature/src/main/scala/lf/typesig/PackageSignature.scala:238]: naive question. Why ?

Comment (simonmaxen-da) [reply daml-lf/api-type-signature/src/main/scala/lf/typesig/PackageSignature.scala:238]: Due to name shadowing the `newSignatures` on line 235 was not the same `newSignatures` as on line 236. Hopefully this makes that more obvious.

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks a lot. 
Maybe worthwhile that @raphael-speyer-da take a look at the ledger service.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 2]
```diff
@@ -391,4 +416,42 @@ object Hash {
       .add(keyHash)
       .add(maintainer)
       .build
+
+  sealed trait KeyPackageName {
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:442]: I have the same reasoning ;-)

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 3]
```diff
@@ -391,4 +416,42 @@ object Hash {
...
+          UsePackageName(name)
+        case _ if version < LanguageVersion.Features.packageUpgrades => NoPackageName
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Invalid combination of package name and language version"
+          )
+      }
+
+    /** Only use in testing */
+    def unsafe(packageName: Option[PackageName]): KeyPackageName = packageName match {
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:473]: Removed.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    ): Node.Action = node match {
+      case n: Node.Create => CreateUpdater(n, version, packageName)
+      case n: Node.Fetch => FetchUpdater(n, version, packageName)
+      case n: Node.Exercise => ExerciseUpdater(n, version, packageName)
+      case n: Node.LookupByKey => LookupByKeyUpdater(n, version, packageName)
+    }
+
+  }
+
+  implicit class Ops[T <: Node.Action](node: T)(implicit versionUpdater: VersionUpdater[T]) {
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala:99]:
```suggestion
  implicit class Ops[T <: Node.Action](val node: T)(implicit versionUpdater: VersionUpdater[T]) extends AnyVal{
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala [hunk 5]
```diff
@@ -298,7 +302,7 @@ object Hash {
   // This function assumes that key is well typed, i.e. :
   // 1 - `templateId` is the identifier for a template with a key of type τ
   // 2 - `key` is a value of type τ
-  @throws[HashingError]
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala:301]: fair enough.

**Review by remyhaemmerle-da — APPROVED** [review]

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala (hunk 4 above)
Comment (simonmaxen-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala:99]: Noted.

File: daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import com.daml.lf.transaction.{GlobalKeyWithMaintainers, Node, TransactionVersion}
+
+import Ordering.Implicits._
+
+object NodeVersionUpdater {
+
+  private def defaultUpgradePkgName: Ref.PackageName =
+    Ref.PackageName.assertFromString("updater-package-name")
+
+  trait VersionUpdater[T] {
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/NodeVersionUpdater.scala:17]: Noted.

**Pull request merged**